Add Optional 128-bit Integer Support to C++ Runtime

AaronWebster opened the pull request:
Adds optional support for 128-bit `UInt` and `Int` fields (65–128 bits) to the C++ backend and runtime, using the compiler's native `__int128_t` / `__uint128_t` where available.

## What this does
- `prelude.emb`: `UInt` and `Int` now permit static sizes up to 128 bits (previously 64).
- The runtime detects 128-bit support via `EMBOSS_HAS_INT128` (derived from `__SIZEOF_INT128__`). When it is unavailable, fields wider than 64 bits fail to compile, exactly as before.
- The concrete 128-bit type is supplied through the `EMBOSS_INT128_T` / `EMBOSS_UINT128_T` macros, which a client may override (e.g. to a library 128-bit type) before including the runtime.
- `LeastWidthInteger`, `BitBlock`, `MemoryAccessor`, and `ByteSwap` are extended to 128 bits. Reads and writes work for full-width, non-full-width (e.g. 72/96-bit), little- and big-endian, arrays, and the text format.

## Scope / limitations
- Anonymous `bits` types remain limited to 64 bits (widening them would require further work in `OffsetBitBlock` and the constraints checker).
- 128-bit support requires a platform/compiler with `__int128` (64-bit GCC/Clang). 32-bit targets and MSVC are unaffected and continue to reject >64-bit fields at compile time.

## Testing
- Runtime unit tests for `ByteSwap`, `MaskToNBits`, `IsPowerOfTwo`, `LeastWidthInteger`, `MemoryAccessor` (LE/BE, non-full-width), and `UIntView` / `IntView` read/write/sign-extension at 128 bits.
- Generated-code integration test (`int128_sizes.emb`) covering little-/big-endian structs, arrays, min/max, `CopyFrom`, out-of-range rejection, and text round-trips.

## Changes since the previous review
This revision rebases onto current `master` and fixes correctness issues in the write/validation path that the earlier version did not handle:

- **Range-check truncation.** `UIntView` / `IntView::CouldWriteValue` compared the candidate value against the field bounds after casting it to `uint64_t` / `int64_t`. With a 128-bit `ValueType` this truncated the value, so an out-of-range value for a 65–127-bit field was accepted, and `TryToWrite` then aborted (or silently truncated) instead of returning `false`. The comparison is now done in the widest supported integer type.
- **`is_signed` vs `__int128`.** `std::is_signed<__int128_t>` is `false` under a strict `-std=c++NN`, so the signed lower-bound check (and negative-value text parsing) was compiled out for 128-bit signed fields. Signedness is now determined via `std::numeric_limits<>::is_signed`, consistent with how these methods are enabled.
- Reverted two unrelated whitespace-only changes flagged in the earlier review, defined the 128-bit `ByteSwap` in terms of `EMBOSS_UINT128_T` for override consistency, and sorted the `testdata/BUILD` entry.

AaronWebster committed:
Add optional 128-bit integer support to C++ runtime

AaronWebster committed:
Add optional 128-bit integer support to C++ runtime

AaronWebster committed:
Refactor: Apply consistent formatting and fix include order

AaronWebster committed:
Refactor: Apply consistent formatting and fix include order

AaronWebster committed:
Merge master into int128-runtime-support

AaronWebster committed:
Update prelude to support 128-bit integers and enhance tests

- Update prelude.emb to allow UInt and Int sizes up to 128 bits
  (previously limited to 64 bits). Added documentation noting that
  128-bit support requires platform support (__uint128_t/__int128_t).

- Update constraints_test.py to test with sizes > 128 bits now that
  128-bit fields are valid.

- Add comprehensive unit tests for 128-bit integer support:
  - emboss_bit_util_test.cc: ByteSwap, MaskToNBits, IsPowerOfTwo tests
  - emboss_memory_util_test.cc: MemoryAccessor, BitBlock tests for
    128-bit values including non-full-width (72, 96 bits) and
    cross-64-bit-boundary operations
  - emboss_prelude_test.cc: UIntView and IntView tests for 128-bit
    read/write, sign extension, and CouldWriteValue

- Add int128_sizes_test.cc integration test for generated code with
  128-bit integer fields (UInt128Sizes, Int128Sizes, big-endian
  variants, and arrays).

- Update int128_sizes.emb documentation and remove AnonymousBits128
  struct (128-bit bits types require separate implementation work).

AaronWebster committed:
Update prelude to support 128-bit integers and enhance tests

- Update prelude.emb to allow UInt and Int sizes up to 128 bits
  (previously limited to 64 bits). Added documentation noting that
  128-bit support requires platform support (__uint128_t/__int128_t).

- Update constraints_test.py to test with sizes > 128 bits now that
  128-bit fields are valid.

- Add comprehensive unit tests for 128-bit integer support:
  - emboss_bit_util_test.cc: ByteSwap, MaskToNBits, IsPowerOfTwo tests
  - emboss_memory_util_test.cc: MemoryAccessor, BitBlock tests for
    128-bit values including non-full-width (72, 96 bits) and
    cross-64-bit-boundary operations
  - emboss_prelude_test.cc: UIntView and IntView tests for 128-bit
    read/write, sign extension, and CouldWriteValue

- Add int128_sizes_test.cc integration test for generated code with
  128-bit integer fields (UInt128Sizes, Int128Sizes, big-endian
  variants, and arrays).

- Update int128_sizes.emb documentation and remove AnonymousBits128
  struct (128-bit bits types require separate implementation work).

jasongraffius commented on runtime/cpp/emboss_defines.h:
```diff
@@ -311,4 +311,14 @@
 
+// Detect 128-bit integer support.
+// __uint128_t and __int128_t are available on GCC and Clang on 64-bit targets.
+#if !defined(EMBOSS_HAS_INT128)
```
For extensibility, we could define `EMBOSS_INT128_T` and `EMBOSS_UINT128_T` type macros defined to be `__int128_t` and `__uint128_t` if not already defined and we have `__SIZEOF_INT128__`. That way client code can supply or override these to enable 128-bit integer support with a custom type or a library type not called `__int128_t` (for example, `absl::int128`).

jasongraffius commented on runtime/cpp/emboss_defines.h:
```diff
@@ -245,7 +245,7 @@
 #if !defined(EMBOSS_ALIAS_SAFE_POINTER_CAST)
 #define EMBOSS_ALIAS_SAFE_POINTER_CAST(t, x) \
-  reinterpret_cast<t __attribute__((__may_alias__)) *>((x))
+  reinterpret_cast<t __attribute__((__may_alias__))*>((x))
```
This seems unrelated? Should probably revert to not pollute the diff.

jasongraffius commented on runtime/cpp/emboss_cpp_types.h:
```diff
@@ -21,6 +21,8 @@
 #include <cstdint>
 #include <type_traits>
 
+#include "runtime/cpp/emboss_defines.h"
```
I think the relative import here should just be "emboss_defines.h" since it's in the same directory.

jasongraffius commented on compiler/back_end/cpp/testcode/int128_sizes_test.cc:
```diff
@@ -0,0 +1,342 @@
+// Copyright 2024 Google LLC
```
2024 -> 2026

jasongraffius commented on testdata/__init__.py:
```diff
@@ -1,13 +0,0 @@
-# Copyright 2019 Google LLC
```
I don't think we want to delete this copyright header

jasongraffius commented on testdata/int128_sizes.emb:
```diff
@@ -0,0 +1,85 @@
+# Copyright 2024 Google LLC
```
2024 -> 2026

AaronWebster committed:
Address PR feedback for 128-bit integer support

- Add EMBOSS_INT128_T and EMBOSS_UINT128_T macros for extensibility.
- Change include path in emboss_cpp_types.h to relative.
- Update copyright years to 2026 in testcode and .emb files.
- Restore testdata/__init__.py copyright header.
- Revert unrelated whitespace change.

AaronWebster committed:
Restore testdata/__init__.py copyright header

AaronWebster commented:
@jasongraffius Ready for another look

AaronWebster committed:
Fix range check for integer fields wider than 64 bits

UIntView::CouldWriteValue and IntView::CouldWriteValue range-checked the
candidate value after casting it to uint64_t / int64_t.  Once fields wider
than 64 bits are allowed, the view's ValueType is __int128 and those casts
truncate the value before it is compared against the field bounds, so an
out-of-range value whose low 64 bits happen to fall in range is accepted.
TryToWrite then proceeds into WriteUInt, which either aborts (EMBOSS_CHECK)
or silently truncates -- so the "safe" write path crashes on out-of-range
input instead of returning false.

Additionally, IntView::CouldWriteValue gated its lower-bound check on
::std::is_signed<IntT>, which is false for __int128_t under a strict
-std=c++NN (extended integer types are not "integral types" in the standard
sense).  The method is *enabled* via ::std::numeric_limits<>::is_integer, so
the signedness test is now made the same way, via
::std::numeric_limits<>::is_signed, keeping the two consistent.

Compare against the field bounds in the widest supported integer type
(MaxWidthUnsigned / MaxWidthSigned) so no truncation occurs.  Add regression
tests at a non-power-of-two width above 64 bits (72) covering both the
CouldWriteValue result and that TryToWrite fails gracefully rather than
aborting.

AaronWebster committed:
Address review feedback: revert unrelated changes and tidy 128-bit support

- emboss_defines.h: revert the incidental EMBOSS_ALIAS_SAFE_POINTER_CAST
  whitespace change that was flagged in review as unrelated to this change.
- parser_types.py: revert an unrelated trailing-whitespace-only edit.
- emboss_bit_util.h: define the 128-bit ByteSwap overload in terms of
  EMBOSS_UINT128_T rather than the raw __uint128_t, so that a client that
  overrides the type macro (e.g. to a library 128-bit type) also gets a
  consistent ByteSwap, matching how the type is used elsewhere.
- testdata/BUILD: sort int128_sizes.emb before int_sizes.emb (buildifier).

AaronWebster committed:
Support text format for integer fields wider than 64 bits

DecodeInteger gated its handling of a leading '-' on ::std::is_signed<IntType>,
which is false for __int128_t under a strict -std=c++NN.  As a result a
negative value could not be parsed into a signed field wider than 64 bits (the
'-' was treated as an invalid character and parsing failed), even though the
same value encodes fine on the way out.  Use ::std::numeric_limits<>::is_signed
instead, matching the fix applied to IntView::CouldWriteValue.

Add text-format round-trip tests for values that do not fit in 64 bits,
including a negative value into a >64-bit signed field and rejection of an
out-of-range text value.

robrussell commented on compiler/front_end/constraints_test.py:
```diff
@@ -235,9 +235,10 @@ def test_struct_field_too_big_for_type(self):
         )
 
     def test_bits_field_too_big_for_type(self):
+        # UInt now supports up to 128 bits, so test with 129 bits (17 bytes)
```
nit: "now" won't be relevant after this PR is in - so "UInt supports...", same below.

robrussell commented on testdata/int128_sizes.emb:
```diff
@@ -0,0 +1,85 @@
+  168 [+32]  UInt:128[2]  sixteen_byte
+
+
+# Note: Anonymous `bits` types are still constrained to 64 bits maximum.
```
This comment would be more useful in the docs, maybe in the roadmap.

robrussell commented on compiler/back_end/cpp/testcode/int128_sizes_test.cc:
```diff
@@ -0,0 +1,426 @@
+  EXPECT_TRUE(::emboss::UpdateFromText(writer, "{nine_byte: -12345}"));
+  EXPECT_EQ(static_cast<__int128_t>(-12345), writer.nine_byte().Read());
+
+  // A large negative 128-bit value (-2**120) does not fit in 64 bits; it must
```
(-(2**120))

robrussell commented on compiler/back_end/cpp/testcode/int128_sizes_test.cc:
```diff
@@ -0,0 +1,426 @@
+  // A large negative 128-bit value (-2**120) does not fit in 64 bits; it must
+  // round-trip through the decimal text format.
+  EXPECT_TRUE(::emboss::UpdateFromText(
+      writer, "{sixteen_byte: -1329227995784915872903807060280344576}"));
```
If this value were in hex it would be easier to relate back to the comment above.

robrussell commented on compiler/front_end/prelude.emb:
```diff
@@ -25,14 +25,24 @@
+  -- 128-bit integers (__uint128_t). The C++ backend will use EMBOSS_HAS_INT128
+  -- to detect availability. On platforms without 128-bit integer support,
+  -- fields larger than 64 bits will fail to compile.
+  [static_requirements: $is_statically_sized && 1 <= $static_size_in_bits <= 128]
```
It looks like there's a place where a check for this this could be included at https://github.com/google/emboss/blob/b9f89e3572f815eb141782f102fcfa0bba6a034c/compiler/front_end/constraints.py#L704 but I'm not sure constraints.py would have enough information to know whether the platform supports 128-bit integers.

robrussell commented on runtime/cpp/emboss_cpp_types.h:
```diff
@@ -51,12 +53,26 @@ struct FloatType<32> final {
 template <int kBits>
 struct LeastWidthInteger final {
   static_assert(kBits <= 64, "Only bit sizes up to 64 are supported.");
   using Unsigned = typename LeastWidthInteger<kBits + 1>::Unsigned;
```
I'm a little confused about this (including the previously existing implementation)
If `kBits` is 64 then this is 

```
using Unsigned = typename LeastWidthInteger<65>::Unsigned;
```

Is this `using` just never applicable because the specializations for 64, 32, 16, and 8 always apply? Or is there something else I'm missing?

robrussell reviewed (changes requested):
The docs at https://github.com/google/emboss/blob/master/doc/language-reference.md#uint still say max size of `Int` and `UInt` are both 64 bits.

The different limit for the size of anonymous bit fields should be mentioned at https://github.com/google/emboss/blob/master/doc/language-reference.md#anonymous-bits .

AaronWebster committed:
Address review feedback: update docs and clarify comments for 128-bit support

- Document the new 128-bit UInt/Int limits (and the platform-support
  caveat) in the language reference, and note the unchanged 64-bit limit
  on bits types, including anonymous bits.
- Move the bits-size note out of testdata/int128_sizes.emb into the docs.
- Reword test comments (drop "now"; -(2**120) instead of ambiguous
  -2**120) and use a hex literal in the text-format test so the value is
  easy to relate to the expected shifted constant.
- Explain the LeastWidthInteger rounding-up recursion in a comment and
  make the no-int128 static_assert message say why the width is
  unsupported.
